Add pool based memory resource

martinus opened the pull request:
A memory resource similar to `std::pmr::unsynchronized_pool_resource`, but optimized for node-based containers. The goal is to be able to cache more coins with the same memory usage, and allocate/deallocate faster.

This is a reimplementation of #22702. The goal was to implement it in a way that is simpler to review & test

* There is now a generic `PoolResource` for allocating/deallocating memory. This has practically the same API as `std::pmr::memory_resource`. (Unfortunately I cannot use std::pmr because libc++ simply doesn't implement that API).
* Thanks to sipa there is now a fuzzer for PoolResource! On a fast machine I ran it for ~770 million executions without finding any issue.

* The estimation of the correct node size is now gone, PoolResource now has multiple pools and just needs to be created large enough to have space for the unordered_map nodes.

I ran benchmarks with #22702, mergebase, and this PR. Frequency locked Intel i7-8700, clang++ 13.0.1 to reindex up to block 690000.

```sh
bitcoind -dbcache=5000 -assumevalid=00000000000000000002a23d6df20eecec15b21d32c75833cce28f113de888b7 -reindex-chainstate -printtoconsole=0 -stopatheight=690000
```

The performance is practically identical with #22702, just 0.4% slower. It's ~21% faster than master:

![Progress in Million Transactions over Time(2)](https://user-images.githubusercontent.com/14386/173288685-91952ade-f304-4825-8bfb-0725a71ca17b.png)

![Size of Cache in MiB over Time](https://user-images.githubusercontent.com/14386/173291421-e6b410be-ac77-479b-ad24-5fafcebf81eb.png)
Note that on cache drops mergebase's memory doesnt go so far down because it does not free the `CCoinsMap` bucket array.

![Size of Cache in Million tx over Time(1)](https://user-images.githubusercontent.com/14386/173288703-a80c9c9e-93c8-4a16-9df8-610c89c61cc4.png)

martinus committed:
Add pool based memory resource & allocator

A memory resource similar to std::pmr::unsynchronized_pool_resource, but
optimized for node-based containers.

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

martinus committed:
Calculate memory usage correctly for unordered_maps that use PoolAllocator

Extracts the resource from a PoolAllocator and uses it for
calculation of the node's memory usage.

martinus committed:
Add PoolResource fuzzer

Fuzzes PoolResource with random allocations/deallocations, and multiple
asserts.

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

martinus committed:
Call ReallocateCache() on each Flush()

This frees up all associated memory with the map, not only the nodes.
This is necessary in preparation for using the PoolAllocator for
CCoinsMap, which does not actually free any memory on clear().

martinus committed:
Use PoolAllocator for CCoinsMap

In my benchmarks, using this pool allocator for CCoinsMap gives about
20% faster `-reindex-chainstate` with -dbcache=5000 with practically the
same memory usage. The change in max RSS changed was 0.3%.

The `validation_flush_tests` tests need to be updated because
memory allocation is now done in large pools instead of one node at a
time, so the limits need to be updated accordingly.

martinus commented:
rebased to 9205b60 with minor fixes in `pool.h` so it is usable in `boost::unordered_map`

martinus commented on src/test/util/xoroshiro128plusplus.h:
```diff
@@ -0,0 +1,72 @@
+        return (x << n) | (x >> (64 - n));
+    }
+
+    [[nodiscard]] constexpr static uint64_t SplitMix64(uint64_t& seedval) noexcept
```
Actually this it is necessary because `seedval` is modified, see the constructor in line 42 where `m_s0` and `m_s1` are initialized. I could use a pointer instead of reference or add a comment to make that more clear?

martinus commented on src/test/util/xoroshiro128plusplus.h:
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
```
That's how rumors start! Maybe it was Satoshi? ;-) I'll update that.

martinus replied:
done

martinus commented:
In a semi-related event, I've recently created a big benchmark of a lot of different hashmaps. See here: https://martin.ankerl.com/2022/08/27/hashmap-bench-01/

Some findings that I think are relevant for this PR:

* I have added the `PoolAllocator` developed here to the benchmarks and combined it with both `std::unordered_map` and `boost::unordered_map` and it works without any problems.
* boost has completely reworked the hashmap and released that in version 1.80. Just changing the map to that implementation will probably provide some speedup. [More info here](https://www.boost.org/doc/libs/1_80_0/libs/unordered/doc/html/unordered.html#benchmarks).
* Both `boost::unordered_map` and `std::unordered_map` benefit a lot memory-wise from the `PoolAllocator`.
* Boost has it's own pool allocator: `boost::container::pmr::unsynchronized_pool_resource` It behaves relatively similar to PoolAllocator, although slower in my benchmarks. So this might be an alternative to the `PoolAllocator`.
* Other implementations like `gtl::parallel_flat_hash_map` are much faster and have low memory overhead, so they might be a good choice performance wise, although they carry much more risk because usage is not so widespread.

martinus commented:
Rebased to [f1b20f0](https://github.com/bitcoin/bitcoin/commit/f1b20f04644ddc07bd9bbfc066f9701d54139d52) to fix copyright header in `xoroshiro128plusplus.h`. Nothing else changed.

martinus commented:
Rebase fixes merge conflict in `Makefile.test_util.include`, fixes a comment, and adds benchmark priority.

aureleoules commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,352 @@
+    /**
+     * Contains all allocated pools of memory, used to free the data in the destructor.
+     */
+    std::list<std::byte*> m_allocated_chunks{};
```
de694ed10bfa623596a41e35c95c13e757787b07: might slightly improve performance but haven't benchmarked
```suggestion
    std::vector<std::byte*> m_allocated_chunks{};
```

martinus replied:
I thought std::list is more fitting here, because I don't need random access, and only every once in a while one element is added. The list is only iterated once when bitcoind shutsdown. So it's not at all performance relevant here

aureleoules reviewed (approved):
The code looks good but it is spooky to review.
I have benchmarked the PR by running `bitcoind -dbcache=5000 -assumevalid=00000000000000000002a23d6df20eecec15b21d32c75833cce28f113de888b7 -reindex-chainstate -printtoconsole=0` and compared the time elapsed between the reindexing of blocks between height=165618 and height=686269 on both master and this PR on mainnet.

Master took 18h06mins and this PR took 13h46mins, corresponding to a reduction of ~24% which matches the PR description.

aureleoules reviewed (commented):
I have also verified that using the default cache (`450MB`), this PR reindexes the chainstate faster than on master.

### Master
`./src/bitcoind -testnet -reindex-chainstate  -stopatheight=1000000 -printtoconsole=0`
real	10m6.978s
user	9m4.140s
sys	1m34.479s

### PR
`./src/bitcoind -testnet -reindex-chainstate  -stopatheight=1000000 -printtoconsole=0`
real	9m32.270s
user	8m29.061s
sys	1m26.911s

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,352 @@
+    /**
+     * Replaces node with placement constructed ListNode that points to the previous node
+     */
+    void PlacementAddToList(void* p, ListNode*& node)
```
I think it's better to leave this as void*, because I don't want any overloaded operator new or operator delete called depending on the type.

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,352 @@
+ * end, a new chunk will be allocated and added to the list.
+ */
+template <std::size_t MAX_BLOCK_SIZE_BYTES, std::size_t ALIGN_BYTES>
+class PoolResource final
```
I'll add these when I rebase to fix the merge conflict :+1:

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+     * Single linked lists of all data that came from deallocating.
+     * m_free_lists[n] will serve blocks of size n*BLOCK_ALIGNMENT_BYTES.
+     */
+    std::array<ListNode*, MAX_BLOCK_SIZE_BYTES / ELEM_ALIGN_BYTES + 1> m_free_lists{};
```
I'm not 100% sure that is the case

In https://eel.is/c++draft/allocator.requirements#general-39 it states that "the return value is unspecified", so it it seems to me that it can be any value, but it doesn't say that the behavior is unspecified. So it could return any pointer or even `nullptr`

In http://eel.is/c++draft/basic.stc.dynamic.allocation#2 it says "The effect of indirecting through a pointer returned from a request for zero size is undefined", so one might return any value, but it must not be dereferenced (which seems logic, with 0 elements there's nothing to dereference). The footnote here says "C++ differs from C in requiring a zero request to return a non-null pointer", so this too seems like a pointer should be returned

So I think `allocate(0)` is not forbidden

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+        if (IsFreeListUsable(bytes, alignment)) {
+            const std::size_t num_alignments = NumElemAlignBytes(bytes);
+
+            // If alignment is lower than BLOCK_ALIGNMENT_BYTES, we need to allocate a bit more.
```
I'll remove the comment, it doesn't make much sense

martinus replied:
done in rebase to 80722d8

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+                return std::exchange(m_free_lists[num_alignments], m_free_lists[num_alignments]->m_next);
+            }
+
+            // freelist is empty: get one allocation from allocated chunk memory.
```
The location seems ok to me, when that line is reached we know the freelist is empty so we need to get memory from the chunk

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+                AllocateChunk();
+            }
+
+            // Make sure we use the right amount of bytes for that freelist (might be rounded up),
```
I think that line fits too, this comments on  the `m_available_memory_it + round_bytes`

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+    /**
+     * Contains all allocated pools of memory, used to free the data in the destructor.
+     */
+    std::list<std::byte*> m_allocated_chunks{};
```
I could, but itt would be less convenient, because forward_list doesn't even have a `size()` which I use in `NumAllocatedChunks()`

jonatack commented on src/Makefile.test.include:
```diff
@@ -112,6 +112,7 @@ BITCOIN_TESTS =\
   test/net_peer_eviction_tests.cpp \
   test/net_tests.cpp \
   test/netbase_tests.cpp \
+  test/pool_tests.cpp \
```
Noticed while running `make check`: the following minor change in the first commit  2bbe0f8eeae274fc30d924c6762cbb7a392f3b32 will run the tests in the expected order.

```diff
   test/net_tests.cpp \
   test/netbase_tests.cpp \
-  test/pool_tests.cpp \
   test/orphanage_tests.cpp \
   test/pmt_tests.cpp \
   test/policy_fee_tests.cpp \
   test/policyestimator_tests.cpp \
+  test/pool_tests.cpp \
   test/pow_tests.cpp \
```

currently

```
Running tests: net_peer_eviction_tests from test/net_peer_eviction_tests.cpp
Running tests: net_tests from test/net_tests.cpp
Running tests: netbase_tests from test/netbase_tests.cpp
Running tests: pool_tests from test/pool_tests.cpp
Running tests: orphanage_tests from test/orphanage_tests.cpp
Running tests: pmt_tests from test/pmt_tests.cpp
Running tests: policy_fee_tests from test/policy_fee_tests.cpp
```

after

```
Running tests: pmt_tests from test/pmt_tests.cpp
Running tests: policy_fee_tests from test/policy_fee_tests.cpp
Running tests: policyestimator_tests from test/policyestimator_tests.cpp
Running tests: pool_tests from test/pool_tests.cpp
Running tests: pow_tests from test/pow_tests.cpp
Running tests: prevector_tests from test/prevector_tests.cpp
Running tests: raii_event_tests from test/raii_event_tests.cpp
```

martinus replied:
Interesting that the order here has any effect on the order of tests, I would have expected that the test suite orders alphabetically. I'll add this to my next update

martinus replied:
done in e7158613dcafd0065a94b03c8013ee4ced8ec3e3

jonatack reviewed (commented):
Reviewing. Verified that each commit @  80722d838f9a5fdbbbd82f97d353e45b5d375ad8 builds cleanly with unit tests green.

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+ *        └───┘
+ *
+ * Here m_free_lists[1] holds the 2 blocks of size 8 bytes, and m_free_lists[2]
+ * holds the3 blocks of size 16. The blocks came from the data stored in the
```
done in rebase to 80722d8

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+            return std::exchange(m_available_memory_it, m_available_memory_it + round_bytes);
+        }
+
+        // Can't use the pool => forward allocation to the upstream resource.
```
done in rebase to 80722d8

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    /**
+     * Returns a block to the freelists, or deletes the block when it did not come from the cunks.
```
done in rebase to 80722d8

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+            // into the memory since we can be sure the alignment is correct.
+            PlacementAddToList(p, m_free_lists[num_alignments]);
+        } else {
+            // Can't use the pool => forward deallocation to the upstream resource.
```
done in rebase to 80722d8

martinus commented on src/coins.h:
```diff
@@ -217,7 +226,8 @@ class CCoinsViewCache : public CCoinsViewBacked
      */
     mutable uint256 hashBlock;
-    mutable CCoinsMap cacheCoins;
+    mutable CCoinsMapMemoryResource cacheCoinsMemoryResource{};
```
done in rebase to 80722d8

martinus commented on src/test/util/poolresourcetester.h:
```diff
@@ -0,0 +1,127 @@
+                chunk_ptr_remaining = chunk_it->ptr;
+                chunk_size_remaining = chunk_it->size;
+            }
+            // std::cout << "free_block=(" << (void*)free_block.ptr << ", " << free_block.size << "), chunk=(" << (void*)chunk_ptr_remaining << ", " << chunk_size_remaining << ")" << std::endl;
```
removed in rebase to 80722d838f9a5fdbbbd82f97d353e45b5d375ad8

sipa reviewed (commented):
Code review ACK apart from one comment:

jonatack reviewed (commented):
WIP review of the first commit 2bbe0f8eeae274fc30, none of my comments below are blockers.

<details><summary>Some minor testing and poking around</summary><p>

Have been running this patch on mainnet since January 20 without issues (M1 macOS 13.1/13.2, clang 15) but not yet done the reindex described the OP.

Bench results of `$ ./src/bench/bench_bitcoin -filter='Pool*.*'`
```diff
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               29.85 |       33,505,082.42 |    0.3% |      0.02 | `PoolAllocator_StdUnorderedMap`
|               13.03 |       76,732,499.95 |    0.4% |      0.01 | `PoolAllocator_StdUnorderedMapWithPoolResource`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               28.63 |       34,927,088.12 |    0.2% |      0.02 | `PoolAllocator_StdUnorderedMap`
|               12.56 |       79,586,875.90 |    0.5% |      0.01 | `PoolAllocator_StdUnorderedMapWithPoolResource`
```

These are the values that build on my machine (fuzz test disabled):

```diff
     static_assert((ELEM_ALIGN_BYTES & (ELEM_ALIGN_BYTES - 1)) == 0, "ELEM_ALIGN_BYTES must be a power of two");
     static_assert(sizeof(ListNode) <= ELEM_ALIGN_BYTES, "Units of size ELEM_SIZE_ALIGN need to be able to store a ListNode");
     static_assert((MAX_BLOCK_SIZE_BYTES & (ELEM_ALIGN_BYTES - 1)) == 0, "MAX_BLOCK_SIZE_BYTES needs to be a multiple of the alignment.");
+    static_assert(ALIGN_BYTES == 8);
+    static_assert(sizeof(ListNode) == 8);
+    static_assert(alignof(ListNode) == 8);
+    static_assert(ELEM_ALIGN_BYTES == 8);
```

</p></details>

jonatack commented on src/test/pool_tests.cpp:
```diff
@@ -0,0 +1,176 @@
+        }
+
+        // Eventually the resource_map should have a much lower memory usage because it has less malloc overhead
+        BOOST_TEST(memusage::DynamicUsage(resource_map) <= memusage::DynamicUsage(std_map) * 90 / 100);
```
Nice. (In practice on my particular system, `86 / 100` is the lowest value that passes.)

martinus replied:
I think the savings are a bit less than 86/100 on 32bit due to 32bit alignment, that's why I have the number a bit on the high side here

jonatack reviewed (commented):
ACK 80722d838f9a5fdbbbd82f97d353e45b5d375ad8 modulo open comments and questions

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+
+            // freelist is empty: get one allocation from allocated chunk memory.
+            const size_t round_bytes = num_alignments * ELEM_ALIGN_BYTES;
+            if (m_available_memory_it + round_bytes > m_available_memory_end) {
```
nice find, I'll change the `if`

martinus replied:
done in e7158613dcafd0065a94b03c8013ee4ced8ec3e3

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+    bool IsEqual(const PoolResource& other) const noexcept
+    {
+        return this == &other;
+    }
```
Right, this is unused; I originally used this in the `operator==`. I'll remove it

martinus replied:
done in e7158613dcafd0065a94b03c8013ee4ced8ec3e3

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+    /**
+     * Points to the end of available memory for carving out allocations.
+     *
+     * That member variable is redundant, and is always equal to `m_allocated_chunks.back().get() + CHUNK_SIZE_BYTES`
```
right, also the .get() is no more

martinus replied:
done in e7158613dcafd0065a94b03c8013ee4ced8ec3e3

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+
+    /**
+     * Single linked lists of all data that came from deallocating.
+     * m_free_lists[n] will serve blocks of size n*BLOCK_ALIGNMENT_BYTES.
```
It should be `n*ELEM_ALIGN_BYTES`

martinus replied:
done in e7158613dcafd0065a94b03c8013ee4ced8ec3e3

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+template <class T1, class T2, std::size_t MAX_BLOCK_SIZE_BYTES, std::size_t ALIGN_BYTES>
+bool operator!=(const PoolAllocator<T1, MAX_BLOCK_SIZE_BYTES, ALIGN_BYTES>& a,
+                const PoolAllocator<T2, MAX_BLOCK_SIZE_BYTES, ALIGN_BYTES>& b) noexcept
+{
```
It's not actually used anywhere currently, but this is needed when move-assigning containers that use the allocator. The containers need to check in a move-assignment if the allocators `operator==` returns true or false, and if they return true it means that both containers use the same resource and they can just destroy lhs and then move the pointers of rhs to lhs without copying elements. When they don't compare equal when the resource is different, all the objects in rhs need to be recreated.

The best explanation that I found for this stuff is this post from Howard Hinnant (he wrote `std::chrono` among other things) : https://stackoverflow.com/a/27472502/48181

jonatack commented:
ACK e7158613dcafd0065a94b03c8013ee4ced8ec3e3 per `git range-diff 4b51290 80722d8 e715861`

modulo any update for https://github.com/bitcoin/bitcoin/pull/25325#discussion_r1086981056, happy to re-review if needed

jonatack commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,358 @@
+    }
+
+    PoolAllocator(const PoolAllocator& other) noexcept = default;
+    PoolAllocator& operator=(const PoolAllocator& other) noexcept = default;
```
Thanks. I wondered if the move constructor is used, because while this branch compiles for me (Clang 15 on ARM64) with the move assignment disabled (`PoolAllocator& operator=(PoolAllocator&&) noexcept = delete;`), it doesn't build with the move constructor disabled (`PoolAllocator(PoolAllocator&&) noexcept = delete;`).

```
./coins.h:230:33: note: in instantiation of member function 'std::unordered_map<COutPoint, CCoinsCacheEntry, SaltedOutpointHasher, std::equal_to<COutPoint>, PoolAllocator<std::pair<const COutPoint, CCoinsCacheEntry>, 128, 8>>::unordered_map' requested here
    mutable CCoinsMap cacheCoins{0, CCoinsMap::hasher{}, CCoinsMap::key_equal{}, &m_cache_coins_memory_resource};
                                ^
./support/allocators/pool.h:297:5: note: 'PoolAllocator' has been explicitly marked deleted here
    PoolAllocator(PoolAllocator&&) noexcept = delete;
    ^
1 error generated.
```

martinus replied:
I think the problem with adding the `PoolAllocator(PoolAllocator&&) noexcept = delete;` is that it makes the function available for overload resolution, but if its chosen the compilation fails https://stackoverflow.com/a/35406415/48181

martinus commented:
Thanks for reviewing @jonatack and @sipa! I've addressed all comments with e715861

martinus commented:
Rebased to resolve conflicts, diff should be viewable with `git range-diff 82903a7 e715861 f58a5eee6f`

jonatack commented:
re-ACK f58a5eee6ff69582d6e46e93dd52e07202d257a8

The unit test CI failures look unrelated -- I didn't reproduce them running `make check` locally (Clang 15 on ARM64).

martinus commented:
Hm it looks like in some environments the new check fails:
```
coins_tests.cpp(984): error: in "coins_tests/ccoins_flush_behavior": check view->DynamicMemoryUsage() < cache_usage has failed
```

jonatack commented:
Fixup LGTM modulo squashing. The remaining CI failures are due to https://github.com/bitcoin/bitcoin/issues/27001.

martinus commented:
Squashed and rebased again due to #26999

See the full range-diff after the 2 rebases: `git range-diff ceb74b8 e715861 c25a754b`

Biggest change was updating the method `TestFlushBehavior` in coins_test.cpp

jonatack commented:
re-ACK 0007d69f249068a14b9b5a97d46ace9dabdc2c8b

martinus commented:
Rebased due to #27011. Note that this adds a `/*deterministic=*/true` [here in test/fuzz/coins_view.cpp](https://github.com/bitcoin/bitcoin/pull/25325/commits/78f597be2879c39d9d2b98e21ed0120d2308de20#diff-1ef3b6a1936b50f3d5ec4a1786d9e2d63d1a3e1815b103e67f20601995f355b4R119), see `git range-diff 0007d69...78f597b`

jonatack reviewed (commented):
re-ACK 78f597be2879c39d9d2b98e21ed0120d2308de20 reviewed `git range-diff 2c1fe27 0007d69 78f597b` and as a sanity check locally ran debug build and unit tests and fuzz build with briefly running the `coins_view` and `coinscache_sim` fuzzers

martinus commented:
yet another rebase, only dropping the "Add xoroshiro128++ PRNG" commit which has already been added in #26153 (commit 5f05b27841af0bed1b6e7de5f46ffe33e5919e4d)

martinus commented:
Rebased adds `include <test/util/random.h>`

john-moffett commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,349 @@
+     * Points to the end of available memory for carving out allocations.
+     *
+     * That member variable is redundant, and is always equal to `m_allocated_chunks.back() + m_chunk_size_bytes`
+     * whenever it is accessed, but `m_untouched_memory_end` caches this for clarity and efficiency.
```
`m_untouched_memory_end` now called `m_available_memory_end`

john-moffett commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,349 @@
+     */
+    void AllocateChunk()
+    {
+        // if there is still any available memory is left, put it into the freelist.
```
Extra "is" after "memory".

john-moffett reviewed (approved):
ACK d87cb99bb37637e26a9e00b9f7de4bc6f44cb79d

Tested and code reviewed. Very nice improvement.

Left tiny nits in documentation, but please ignore unless updating for other reasons.

LarryRuane commented on src/coins.h:
```diff
@@ -131,7 +132,15 @@ struct CCoinsCacheEntry
+                                     SaltedOutpointHasher,
+                                     std::equal_to<COutPoint>,
+                                     PoolAllocator<std::pair<const COutPoint, CCoinsCacheEntry>,
+                                                   sizeof(std::pair<const COutPoint, CCoinsCacheEntry>) + sizeof(void*) * 4,
```
I don't understand the purpose of adding `sizeof(void*) * 4`; could you leave a brief comment if you get the chance? (Unless I'm just being clueless!)

martinus replied:
Posting here now, I can add this as a comment later: 

The value here determines the maximum bytes that the `PoolAllocator` supports. When bigger blocks are allocated, this is just forwarded to `new`.

The thing with `sizeof(void*) * 4` is, it is not enough to just support up to sizes of the `std::pair`, because the different implementations of `std::unordered_map` use more memory for each node. Most implementations wrap the std::pair into a struct that contains a single pointer, so they can link them in a single linked list. But not all; e.g. Microsoft's STL uses a double linked list. Also libstd++ and libc++ wrap the pair differently, which can lead to different memory usage due to alignement.

So with adding 4 * the size of a pointer we err on the safe side, and nodes can surely be allocated with all implementions of std::unordered_map. `m_free_lists` might be a bit larger than it needs to be and have a few pointers that never hold a freelist, but at least we can be sure that `std::unordered_map` nodes are actually allocated with the pool.

hebasto replied:
@martinus

Could assessing of the `std::unordered_map::node_type` improve this heuristic?

martinus replied:
@hebasto Unfortunately not, I actually wanted to use this but learned that this is actually just the type of the node handle which more or less is just a pointer to the data. The actual type of the node used for allocation is not exposed by the standard. If it were, it would make the pool implementation quite a bit simpler

LarryRuane reviewed (commented):
I'd like to review more, but don't let me hold you up. I'm hosting [review club on this PR](https://bitcoincore.reviews/25325) on March 8 (next week); anyone here should feel free to join since you know this PR better than I do!

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,349 @@
+    static constexpr std::size_t ELEM_ALIGN_BYTES = std::max(alignof(ListNode), ALIGN_BYTES);
+    static_assert((ELEM_ALIGN_BYTES & (ELEM_ALIGN_BYTES - 1)) == 0, "ELEM_ALIGN_BYTES must be a power of two");
+    static_assert(sizeof(ListNode) <= ELEM_ALIGN_BYTES, "Units of size ELEM_SIZE_ALIGN need to be able to store a ListNode");
+    static_assert((MAX_BLOCK_SIZE_BYTES & (ELEM_ALIGN_BYTES - 1)) == 0, "MAX_BLOCK_SIZE_BYTES needs to be a multiple of the alignment.");
```
In line 89 there is an assert that `ELEM_ALIGN_BYTES` is multiple of 2:
```cpp
static_assert((ELEM_ALIGN_BYTES & (ELEM_ALIGN_BYTES - 1)) == 0, "ELEM_ALIGN_BYTES must be a power of two");
```
So given that line 91 should make sense, `ELEM_ALIGN_BYTES - 1` becomes a bitmask and actually asserts that `MAX_BLOCK_SIZE_BYTES` is multiple of `ELEM_ALIGN_BYTES`. But right, on its own that assert wouldn't be enough and using `%` is probably a bit more clear.

martinus commented:
@LarryRuane awesome that you'll hold a PR review club about this PR! I'll try to join, but can't yet guarantee that I'll have the time.

LarryRuane commented on src/bench/pool.cpp:
```diff
@@ -0,0 +1,50 @@
+template <typename Map>
+void BenchFillClearMap(benchmark::Bench& bench, Map& map)
+{
+    size_t batch_size = 5000;
+
+    // make sure each iteration of the benchmark contains exactly 5000 inserts and one clear.
+    // do this at least 10 times so we get reasonable accurate results
+
+    bench.batch(batch_size).minEpochIterations(10).run([&] {
+        auto rng = ankerl::nanobench::Rng(1234);
+        for (size_t i = 0; i < batch_size; ++i) {
+            map[rng()];
+        }
+        map.clear();
+    });
```
This benchmark doesn't re-use memory (add to freelist then remove from freelist); maybe it would be better if it did because that's what happens with the real coins cache, maybe something like this:

<details>
<summary>patch</summary>

```diff
-    size_t batch_size = 5000;
+    // The steady-state size of the map will be half of this value;
+    // power-of-2 to avoid expensive mod operation during benchmark.
+    constexpr size_t batch_size = 1 << 13;
 
     // make sure each iteration of the benchmark contains exactly 5000 inserts and one clear.
     // do this at least 10 times so we get reasonable accurate results
 
     bench.batch(batch_size).minEpochIterations(10).run([&] {
         auto rng = ankerl::nanobench::Rng(1234);
-        for (size_t i = 0; i < batch_size; ++i) {
-            map[rng()];
+        for (size_t i = 0; i < batch_size * 10; ++i) {
+            uint64_t r{rng()};
+            // if the map has few entries, more likely add an entry, else delete
+            if ((r & (batch_size-1)) < map.size()) {
+                map.erase(map.begin());
+            } else {
+                map[r];
+            }
         }
         map.clear();
     });
```
</details>

I don't think it matters which entry we delete, since `map.begin()` could return any entry. Deleting a truly random entry gets more complicated, and I don't think it changes the performance significantly. The loop iterates `batch_size * 10` instead of `batch_size` so we reach steady-state (about equal numbers of insertions and deletions).

martinus replied:
Actually it does reuse memory, when calling `map.clear()` the nodes are put back into the freelist. All elements are deallocated, so he `pool_resource` gets them and puts them into their freelist. The memory is only ever released in the destructor of `pool_resource`

LarryRuane replied:
That's a good point, I hadn't noticed that. But perhaps it's good to make the load more realistic to get better benchmarking results. I'm assuming that when UTXOs become spent TXOs during block validation, they're removed from this unordered_map, but maybe that's not true? Just seems like doing many interleaving allocations and deallocations would be more realistic, but maybe not. But this suggestion is non-blocking.

In case anyone is wondering, I observed that the benchmark does cause allocations (and deallocations) that are too large for the pool, and these are for the hash bucket arrays. (I'm sure the fuzzer does that too, so far I've run only the benchmark.)

LarryRuane reviewed (commented):
Looks good.

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,349 @@
+    void AllocateChunk()
+    {
+        // if there is still any available memory is left, put it into the freelist.
+        size_t remaining_available_bytes = std::distance(m_available_memory_it, m_available_memory_end);
```
This is explicitly defined in the standard here, in 5.1: http://eel.is/c++draft/expr.add#5

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,349 @@
+ * some memory available for the blocks, and when m_available_memory_it is at the
+ * end, a new chunk will be allocated and added to the list.
+ */
+template <std::size_t MAX_BLOCK_SIZE_BYTES, std::size_t ALIGN_BYTES>
```
No real reason. Technically in C++ I believe `std::size_t` should be preferred, especially when using includes like `<cstddef>` instead of `<stddef.h>`. I personally usually use just `size_t` everywhere and never had a compile problem.

sipa reviewed (approved):
ACK d87cb99bb37637e26a9e00b9f7de4bc6f44cb79d

jamesob commented:
Kicked off a bitcoinperf run; will have some results tomorrow.

martinus commented:
Thanks for the benchmark @jamesob! There was no cache flush in the benchmark, that's why the memory usage was lower. When flushes would happen with e.g. lower dbcache size or longer range of blocks memory usage should be about equal, but then there should be an even larger performance benefit for this PR because it can cache more data with the same memory.

jamesob commented:
> When flushes would happen with e.g. lower dbcache size or longer range of blocks memory usage should be about equal, but then there should be an even larger performance benefit for this PR because it can cache more data with the same memory.

Yup - I started another run with dbcache=1000.

achow101 commented:
Light ACK d87cb99bb37637e26a9e00b9f7de4bc6f44cb79d

Not particularly well versed in allocators, but the logic of the allocator and its tests makes sense, and the benchmarks seem to show there is a noticeable improvement.

One thing I did notice though is that when configured with `--enable-debug`, the benchmark `PoolAllocator_StdUnorderedMapWithPoolResource` is a little bit slower than `PoolAllocator_StdUnorderedMap`. Without `--enable-debug`, it's quite a bit faster. I didn't measure the effect of this in actual IBD, but configuring with `--enable-debug` is already known to make things slower, so it shouldn't matter much.

With `--enable-debug`:

|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|              239.73 |        4,171,384.12 |    0.1% |        2,516.56 |         401.94 |    0.1% |      0.14 | `PoolAllocator_StdUnorderedMap`
|              277.15 |        3,608,128.61 |    0.1% |        2,792.06 |         432.32 |    0.1% |      0.17 | `PoolAllocator_StdUnorderedMapWithPoolResource`

Without `--enable-debug`:

|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|               21.53 |       46,436,178.12 |    0.3% |          279.84 |          59.34 |    0.4% |      0.01 | `PoolAllocator_StdUnorderedMap`
|               10.45 |       95,674,714.97 |    0.2% |           82.34 |          13.71 |    0.8% |      0.01 | `PoolAllocator_StdUnorderedMapWithPoolResource`

jamesob commented:
Back with results for dbcache=1000; less noticeable speedup (5%) and increased memory usage (11%).

---

![ibd local range dbcache=1000 667200 697200](https://user-images.githubusercontent.com/73197/226771758-dc8f9974-8304-472e-a9fb-15c2e493d5f2.png)

|                 bench name                 |                                                                                                     command                                                                                                     |
|--------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ibd.local.range.dbcache=1000.667200.697200 | `bitcoind -dbcache=1000 -debug=coindb -debug=bench -listen=0 -connect=0 -addnode=127.0.0.1:8888 -prune=9999999 -printtoconsole=0 -assumevalid=000000000000000000176c192f42ad13ab159fdb20198b87e7ba3c001e47b876` |


### #25325 vs. $mergebase (absolute)
|                         bench name                         |  x  |           #25325            |         $mergebase         |
|------------------------------------------------------------|----:|-----------------------------|----------------------------|
| ibd.local.range.dbcache=1000.667200.697200.total_secs      |   3 | 4721.4819 (± 11.2932)       | 4990.6752 (± 19.7466)      |
| ibd.local.range.dbcache=1000.667200.697200.peak_rss_KiB    |   3 | 2906596.0000 (± 86412.3208) | 2607876.0000 (± 4288.9414) |
| ibd.local.range.dbcache=1000.667200.697200.cpu_kernel_secs |   3 | 499.0333 (± 4.3527)         | 581.9333 (± 4.1179)        |
| ibd.local.range.dbcache=1000.667200.697200.cpu_user_secs   |   3 | 26882.3333 (± 17.6349)      | 27490.5000 (± 12.0885)     |


### #25325 vs. $mergebase (relative)
|                         bench name                         |  x  | #25325 | $mergebase |
|------------------------------------------------------------|----:|-------:|-----------:|
| ibd.local.range.dbcache=1000.667200.697200.total_secs      |   3 |   1.00 |      1.057 |
| ibd.local.range.dbcache=1000.667200.697200.peak_rss_KiB    |   3 |   1.11 |      1.000 |
| ibd.local.range.dbcache=1000.667200.697200.cpu_kernel_secs |   3 |   1.00 |      1.166 |
| ibd.local.range.dbcache=1000.667200.697200.cpu_user_secs   |   3 |   1.00 |      1.023 |

LarryRuane commented on src/test/pool_tests.cpp:
```diff
@@ -0,0 +1,177 @@
+            ptr_size_alignment.push_back({ptr, size, alignment});
+        } else {
+            // deallocate a random item
+            auto& x = ptr_size_alignment[InsecureRandRange(ptr_size_alignment.size())];
+            resource.Deallocate(x.ptr, x.bytes, x.alignment);
+            x = ptr_size_alignment.back();
+            ptr_size_alignment.pop_back();
```
I don't understand this, we're not removing the `ptr_size_alignment` entry that we just deallocated; we're removing the last entry. But I don't understand why this test doesn't fail. Also, `x` is unused. Maybe something like:
```suggestion
            auto& x = ptr_size_alignment.back();
            ptr_size_alignment.pop_back();
            resource.Deallocate(x.ptr, x.bytes, x.alignment);
            std::shuffle(ptr_size_alignment.begin(), ptr_size_alignment.end(), ctx);
```

LarryRuane reviewed (commented):
All comments are non-blocking, except I am concerned about zero-length allocations.

I stepped through `pool_tests.cpp` in the debugger and all was as expected. I didn't run or review the fuzz test or the other modified unit tests, but I'll try to get to those in the next couple of days.

martinus commented:
@jamesob interesting that it didn't see a bigger speedup, but I guess it depends on a lot of other factors as well. How fast is your harddisk, and how much RAM does your computer have?

martinus commented on src/test/pool_tests.cpp:
```diff
@@ -0,0 +1,177 @@
+    BOOST_TEST(1 == PoolResourceTester::FreeListSizes(resource)[1]);
+    BOOST_TEST(expected_bytes_available == PoolResourceTester::AvailableMemoryFromChunk(resource));
+
+    // can't use chunk because alignment is too big
+    block = resource.Allocate(8, 16);
```
That here would be `BOOST_TEST(block != b)`, because since `b` now has to come from the `::operator new` and not from the freelist

martinus commented on src/test/pool_tests.cpp:
```diff
@@ -0,0 +1,177 @@
+    BOOST_TEST(1 == PoolResourceTester::FreeListSizes(resource)[1]);
+    BOOST_TEST(expected_bytes_available == PoolResourceTester::AvailableMemoryFromChunk(resource));
+
+    // can't use chunk because size is too big
+    block = resource.Allocate(16, 8);
```
With the above change, if I'm not mistaken I don't think it's safe to check for `BOOST_TEST(block != b)` here too. This too calls `::operator new`, and because the previous memory was deallocated already, it might give out the same block of memory, depending on however malloc is implemented

martinus commented on src/test/pool_tests.cpp:
```diff
@@ -0,0 +1,177 @@
+    BOOST_TEST(expected_bytes_available == PoolResourceTester::AvailableMemoryFromChunk(resource));
+}
+
+// Allocates from 0 to n bytes were n > the PoolResource's data, and each should work
```
After having a closer look, you are right that currently allocating & freeing 0 bytes does not work with the resource. But I think I can't allow allocating 0 bytes with the resource at all, because when allocating multple 0 bytes it would always give out the same pointer, and the standard says that this is not allowed: https://cplusplus.github.io/LWG/issue9

So I think the best solution is to just add a `bytes > 0` condition to `IsFreeListUsable`, so that `::operator new` will be used in that case.

LarryRuane replied:
@martinus, good catch!
@sipa, good idea, this does seem to work (hope this is what you meant):
```
     [[nodiscard]] static constexpr std::size_t NumElemAlignBytes(std::size_t bytes)
     {
-        return (bytes + ELEM_ALIGN_BYTES - 1) / ELEM_ALIGN_BYTES;
+        return bytes > 0 ? ((bytes + ELEM_ALIGN_BYTES - 1) / ELEM_ALIGN_BYTES) : 1;
     }
```
Benchmark results are identical for me, but this should be confirmed.

LarryRuane replied:
That's clever! They generate similar code, but this first way is slightly less code https://godbolt.org/z/Pz77TTEd8
```
int f(int bytes)
{
    return bytes > 0 ? (bytes+8-1)/8 : 1;
}
```

Or https://godbolt.org/z/adzasoPad
```
int f(int bytes)
{
    return (bytes+8-1)/8 + (bytes == 0);
}
```
That's using x86-64 gcc 11.3; results are similar for x86-64 clang 16.0.0.

martinus replied:
The code generation is a bit different when you use unsigned types, then the `+ (bytes == 0)` version is the shortedst for me. In my microbenchmark the `+ (bytes == 0)` is also fastest, but in practice its most likely irrelevant

martinus replied:
I've implemented the `+ (bytes == 0)` version in 9f947fc3d4b779f017332135323b34e8f216f613

martinus commented on src/support/allocators/pool.h:
```diff
@@ -0,0 +1,349 @@
+    void Deallocate(void* p, std::size_t bytes, std::size_t alignment) noexcept
+    {
+        if (IsFreeListUsable(bytes, alignment)) {
+            const std::size_t num_alignments = NumElemAlignBytes(bytes);
+            // put the memory block into the linked list. We can placement construct the FreeList
+            // into the memory since we can be sure the alignment is correct.
+            PlacementAddToList(p, m_free_lists[num_alignments]);
```
I'll implement the ` + (bytes == 0)` for NumElemAlignBytes

martinus commented:
Updated  d87cb99 -> 9f947fc3d4b779f017332135323b34e8f216f613 ([pr25325.1](https://github.com/martinus/bitcoin/commits/pr25325.1) -> [pr25325.2](https://github.com/martinus/bitcoin/commits/pr25325.2))

There is a single behavior change in pool.h, now `NumElemAlignBytes` adds `+ (bytes == 0)` so that allocations of 0 bytes work with the PoolAllocator.

Other than that, updated tests to include allocation of 0 bytes, and fixed all the nits.

martinus commented on src/test/util/poolresourcetester.h:
```diff
@@ -0,0 +1,126 @@
+        for (std::size_t freelist_idx = 0; freelist_idx < resource.m_free_lists.size(); ++freelist_idx) {
+            std::size_t bytes = freelist_idx * resource.ELEM_ALIGN_BYTES;
+            auto* ptr = resource.m_free_lists[freelist_idx];
+            while (ptr != nullptr) {
+                free_blocks.emplace_back(ptr, bytes);
+                ptr = ptr->m_next;
+            }
```
I've left that out on purpose, I didn't want to complicate this code any further. So far this case never happened

john-moffett reviewed (approved):
ACK 9f947fc3d4b779f017332135323b34e8f216f613

Verified the changes with `range-diff`:

```diff
1:  45508ec79 ! 1:  b8401c328 Add pool based memory resource & allocator
    @@ src/support/allocators/pool.h (new)
-     * whenever it is accessed, but `m_untouched_memory_end` caches this for clarity and efficiency.
+     * whenever it is accessed, but `m_available_memory_end` caches this for clarity and efficiency.
    @@ src/support/allocators/pool.h (new)
-     * into m_free_lists.
+     * into m_free_lists. Round up for the special case when bytes==0.
    @@ src/support/allocators/pool.h (new)
-        return (bytes + ELEM_ALIGN_BYTES - 1) / ELEM_ALIGN_BYTES;
+        return (bytes + ELEM_ALIGN_BYTES - 1) / ELEM_ALIGN_BYTES + (bytes == 0);
    @@ src/support/allocators/pool.h (new)
-        // if there is still any available memory is left, put it into the freelist.
+        // if there is still any available memory left, put it into the freelist.
    @@ src/test/pool_tests.cpp (new)
-    block = resource.Allocate(8, 1);
+    void* b = resource.Allocate(8, 1);
+    BOOST_TEST(b == block); // we got the same block of memory as before
    @@ src/test/pool_tests.cpp (new)
-    // can't use chunk because alignment is too big
-    block = resource.Allocate(8, 16);
+    // can't use resource because alignment is too big, allocate system memory
+    b = resource.Allocate(8, 16);
+    BOOST_TEST(b != block);
+    block = b;
    @@ src/test/pool_tests.cpp (new)
+
+    // it's possible that 0 bytes are allocated, make sure this works. In that case the call is forwarded to operator new
+    // 0 bytes takes one entry from the first freelist
+    void* p = resource.Allocate(0, 1);
+    BOOST_TEST(0 == PoolResourceTester::FreeListSizes(resource)[1]);
+    BOOST_TEST(expected_bytes_available == PoolResourceTester::AvailableMemoryFromChunk(resource));
+
+    resource.Deallocate(p, 0, 1);
+    PoolResourceTester::CheckAllDataAccountedFor(resource);
+    BOOST_TEST(1 == PoolResourceTester::FreeListSizes(resource)[1]);
+    BOOST_TEST(expected_bytes_available == PoolResourceTester::AvailableMemoryFromChunk(resource));
    @@ src/test/pool_tests.cpp (new)
-    auto counts = PoolResourceTester::FreeListSizes(resource);
    @@ src/test/pool_tests.cpp (new)
-    for (uint8_t num_bytes = 1; num_bytes < num_allocs; +num_bytes) {
+    for (uint8_t num_bytes = 0; num_bytes < num_allocs; +num_bytes) {
    @@ src/test/pool_tests.cpp (new)
-        // set each byte to i
+        // set each byte to num_bytes
    @@ src/test/pool_tests.cpp (new)
-    uint8_t val = 1;
+    uint8_t val = 0;
    @@ src/test/pool_tests.cpp (new)
-            std::size_t alignment = std::size_t{1} << InsecureRandRange(7);           // 1, 2, ..., 128
-            std::size_t size = (InsecureRandRange(2000) / alignment + 1) * alignment; // multiple of alignment
+            std::size_t alignment = std::size_t{1} << InsecureRandRange(8);          // 1, 2, ..., 128
+            std::size_t size = (InsecureRandRange(200) / alignment + 1) * alignment; // multiple of alignment
    @@ src/test/util/poolresourcetester.h (new)
-        for (std::byte* ptr : resource.m_allocated_chunks) {
+        for (const std::byte* ptr : resource.m_allocated_chunks) {
    @@ src/test/util/poolresourcetester.h (new)
-            assert(free_block.size <= chunk_size_remaining);                 // ensure we no overflow
+            assert(free_block.size <= chunk_size_remaining);                 // ensure no overflow
    @@ src/test/util/poolresourcetester.h (new)
-        assert(chunk_ptr_remaining == chunks.back().ptr + chunks.back().size); // ensure we are t the end of the chunks
+        // ensure we are at the end of the chunks
+        assert(chunk_ptr_remaining == chunk_it->ptr + chunk_it->size);
+    +chunk_it;
+        assert(chunk_it == chunks.end());
2:  ed2e1cbe8 = 2:  e19943f04 Calculate memory usage correctly for unordered_maps that use PoolAllocator
3:  477e16dbf = 3:  1afca6b66 Add PoolResource fuzzer
4:  d6b85474d = 4:  5e4ac5abf Call ReallocateCache() on each Flush()
5:  d87cb99bb ! 5:  9f947fc3d Use PoolAllocator for CCoinsMap
    @@ src/coins.h: struct CCoinsCacheEntry
+/**
+ * PoolAllocator's MAX_BLOCK_SIZE_BYTES parameter here uses sizeof the data, and adds the size
+ * of 4 pointers. We do not know the exact node size used in the std::unordered_node implementation
+ * because it is implementation defined. Most implementations have an overhead of 1 or 2 pointers,
+ * so nodes can be connected in a linked list, and in some cases the hash value is stored as well.
+ * Using an additional sizeof(void*)*4 for MAX_BLOCK_SIZE_BYTES should thus be sufficient so that
+ * all implementations can allocate the nodes from the PoolAllocator.
+ */

```

jonatack reviewed (approved):
Nice review work, @LarryRuane.

re-ACK 9f947fc3d4b779f017332135323b34e8f216f613

LarryRuane commented:
ACK 9f947fc3d4b779f017332135323b34e8f216f613

martinus commented:
@sipa could you have another look after my update from d87cb99 -> 9f947fc3d4b779f017332135323b34e8f216f613?

achow101 commented:
re-ACK 9f947fc3d4b779f017332135323b34e8f216f613

martinus commented:
Wohoo :tada: Thanks everyone for making this happen!
